### PR TITLE
[Feature] Add evil teams assets

### PR DESCRIPTION
**Pull request by Tempo-anon**
## What are the changes?
* Added evil teams to the game assets

## Why am I doing these changes?
* Evil teams are a commonly requested feature
* ~~Classic mode is lacking content in the midgame apart from a few rival battles~~
* ~~Players only need to prepare for the final boss and the rival as the main hurdles in a classic playthrough. This introduces some roguelike randomness where there is another major boss that players have to prepare for~~

## What did change?
* Added sprites and json's for new trainers
* ~~Created an `evilTeam` that is randomly chosen when players begin a new game. This will then spawn grunts that will randomly replace normal trainers with increasing frequency until the evil team boss is defeated at floor 105~~
* ~~Fixed evil team grunt spawns at floors 35. This seed offset will be used to decide what evil team will be seen through the run~~
* ~~Guaranteed evil grunts spawn on floors 61, 63, and 65 (no gym leader on floor 70) as well as 121, and 123 (no gym leader on 130)~~
* ~~Evil team boss spawns on floors 125 and 165~~

## How to test the changes?
Just pull the code and start a classic run. Overrides can be used to skip ahead to the relevant boss fights

## Checklist
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I tested the changes (manually)?
    - [x] Are all unit tests still passing? (`npm run test`)
- [x] Are the changes visual?
  - [x] Have I provided screenshots/videos of the changes?

## Todo
- [x] Dialogue needs to be updated from memes and jokes
- [x] How often grunts spawn probably need to be tweaked
- [ ] Evil Team Pokemon selection and balancing

## Goals
When making the teams of the grunts and bosses I tried to focus on a few principles
* The teams should all be distinct. Try not to reuse Pokemon species that other teams or notable trainers use
* The specific generation/region the team came from should be highlighted. This means team rocket should have an emphasis on gens 1-2 while magma/aqua have a focus on gen 3, etc
* Show off some underutilized features in classic mode such as more mega Pokemon
* Variety and variance among similar Pokemon in a team
* The first boss fight should reference the feel of fighting them in game. The second fight should highlight a legendary Pokemon as well as a more themed team. It _should_ be a challenge!

### CODE REVIEW

File: src/battle-scene.ts [hunk 1]
```diff
@@ -910,7 +978,10 @@ export default class BattleScene extends SceneBase {
       }
 
       if (newBattleType === BattleType.TRAINER) {
-        const trainerType = this.arena.randomTrainerType(newWaveIndex);
+        let trainerType = this.arena.randomTrainerType(newWaveIndex);
+        if (this.evilTeamOverride(waveIndex)) {
+          trainerType = this.getEvilTeamGrunt(this.evilTeamThisRun);
+        }
```
Comment (flx-sta) [inline src/battle-scene.ts:984]: you should not first run `this.arena.randomTrainerType(newWaveIndex);` and then overwrite it after checking `this.evilTeamOverride(waveIndex)`. Instead initialize `trainerType` with the correct one immediately

File: src/battle.ts [hunk 2]
```diff
@@ -326,19 +326,85 @@ interface FixedBattleConfigs {
...
+      return TrainerType.PLASMA_GRUNT;
+    case EvilTeam.TEAM_FLARE:
+      return TrainerType.FLARE_GRUNT;
+    default:
+      return TrainerType.ROCKET_GRUNT;
+    }
+  case "boss1":
+    switch (evilTeam) {
+    case EvilTeam.TEAM_ROCKET:
+      return TrainerType.ROCKET_BOSS_GIOVANNI_1;
```
Comment (flx-sta) [inline src/battle.ts:351]: you can ditch this case (or make it fallthrough) as it returns the same as `default`

File: src/battle.ts [hunk 3]
```diff
@@ -326,19 +326,85 @@ interface FixedBattleConfigs {
...
+      return TrainerType.GHETSIS;
+    case EvilTeam.TEAM_FLARE:
+      return TrainerType.LYSANDRE;
+    default:
+      return TrainerType.ROCKET_BOSS_GIOVANNI_1;
+    }
+  case "boss2":
+    switch (evilTeam) {
+    case EvilTeam.TEAM_ROCKET:
+      return TrainerType.ROCKET_BOSS_GIOVANNI_2;
```
Comment (flx-sta) [inline src/battle.ts:368]: you can ditch this case (or make it fallthrough) as it returns the same as `default`

File: src/locales/de/trainers.ts [hunk 4]
```diff
@@ -116,7 +122,25 @@ export const trainerClasses: SimpleTranslationEntries = {
   "worker": "Arbeiter",
   "worker_female": "Arbeiterin",
   "workers": "Arbeiter",
-  "youngster": "Knirps"
+  "youngster": "Knirps",
+  "rocket_grunt": "Rüpel von Team Rocket",
```
Comment (flx-sta) [inline src/locales/de/trainers.ts:126]: I think it's actually `Team Rocket Rüpel` 
(same for the other following ones)

File: src/battle-scene.ts [hunk 5]
```diff
@@ -880,6 +899,55 @@ export default class BattleScene extends SceneBase {
...
+    }
+    // Evil team always spawns leading up to the boss
+    if (waveIndex > 99 && waveIndex < 105) {
+      return true;
+    }
+    // Evil team spawns scale up
+    if (waveIndex < 100) {
+      return Utils.randInt(10) < Math.floor(waveIndex / 10);
+    }
+  }
```
Comment (flx-sta) [inline src/battle-scene.ts:927]: Maybe these conditions should be wrapped into some kind of methods in a more "global" manner.
Otherwise eventual other changes in the future might not know that these waves are reserved for the evils

**Review by flx-sta — CHANGES_REQUESTED** [review]

File: src/locales/de/trainers.ts (hunk 4 above)
Comment (CodeTappert) [reply src/locales/de/trainers.ts:126]: See https://www.pokewiki.de/R%C3%BCpel_von_Team_Rocket

Comment (CodeTappert) [reply src/locales/de/trainers.ts:126]: Or here for the newest team (star): https://www.pokewiki.de/R%C3%BCpel_von_Team_Star

Comment (flx-sta) [reply src/locales/de/trainers.ts:126]: Even with sauce. Can't beat that

**Comment by SeafoamQueen** [comment]
I was tester for this feature. I used a fire/water/grass team composed of pokemon available at the start of the game. I did use egg moves on them, but I don't feel like this substantially altered the viability of my team.

The new trainer battles felt fair in terms of both frequency and opposing team compositions.
This change does increase the difficulty of classic, but also increases the replayability.
It also increases mid-game engagement for players who are able to consistently complete classic in it's current state.

File: src/battle-scene.ts (hunk 1 above)
Comment (Tempo-anon) [reply src/battle-scene.ts:984]: I don't think it makes a functional difference besides neatness if I made it like this. Having it as separate lines in my mind makes it easier to remove/change since the final formula for overriding hasn't been determined yet
```
let trainerType = this.evilTeamOverride(waveIndex) ? this.getEvilTeamGrunt(this.evilTeamThisRun) : this.arena.randomTrainerType(newWaveIndex)
```

File: src/battle.ts (hunk 2 above)
Comment (Tempo-anon) [reply src/battle.ts:351]: I think this is just a styling preference since I like having explicit cases for each possibility as well as a default in case things go wrong, but you are right that I can skip the EvilTeam.TEAM_ROCKET case if they're also in the default

**Comment by Tempo-anon** [comment]
> It seems like i told you the wrong sprite names yesterday...The game searches for magma_grunt_m and magma_grunt_f. (same for the other teams)
> 
> So maybe you can still revert that change. If not i will create a PR to this anyways since i want to swap the magma grunts to the M ones.
> 
> Here is the PR: [Tempo-anon#2](https://github.com/Tempo-anon/pokerogue/pull/2)

You tricked me 😭 it's fine, I'll fix it in a new commit soon

File: src/battle-scene.ts (hunk 5 above)
Comment (Tempo-anon) [reply src/battle-scene.ts:927]: The only reserved waves for evils right now are defined at the bottom of `battle.ts` which is 35, 111, and 113 right now. This is just a filler for now but I guess I can make it a bit more detailed

File: src/battle.ts (hunk 2 above)
Comment (flx-sta) [reply src/battle.ts:351]: Or use a fall through:
```ts
    switch (evilTeam) {
    case EvilTeam.TEAM_MAGMA:
      return TrainerType.MAGMA_GRUNT;
    case EvilTeam.TEAM_AQUA:
      return TrainerType.AQUA_GRUNT;
    case EvilTeam.TEAM_GALACTIC:
      return TrainerType.GALACTIC_GRUNT;
    case EvilTeam.TEAM_PLASMA:
      return TrainerType.PLASMA_GRUNT;
    case EvilTeam.TEAM_FLARE:
      return TrainerType.FLARE_GRUNT;
    case EvilTeam.TEAM_ROCKET:
    default:
      return TrainerType.ROCKET_GRUNT;
```

File: src/battle-scene.ts (hunk 1 above)
Comment (flx-sta) [reply src/battle-scene.ts:989]: You don't have to use short-if.

```ts
let trainerType: Trainer //idk the type tbh..

if (this.evilTeamOverride(waveIndex)) {
    trainerType = this.getEvilTeamGrunt(this.evilTeamThisRun) 
} else {
    trainerType = this.arena.randomTrainerType(newWaveIndex)
}
```

it saves processing time as the thread doesn't have to initialize two times

**Comment by star-krieg** [comment]
I tested this feature on a fresh save (I used Fuecoco and 2 other starters for coverage - Mudkip and Turtwig). I failed a few runs because of bad RNG, but my best run so far was able to up to 195. The evil team bosses bring more challenge to classic  and reduces the monotony of having only the rival as trainers to actually prepare for during successive runs. 

I had to repeat a battle a few times to beat the 165 evil boss (I had Giovanni), and preparing a well-balanced team certainly helped. While this makes classic a bit more difficult, it provides another challenge to consider for the player.

**Comment by CodeTappert** [comment]
German Dialogue: https://github.com/Tempo-anon/pokerogue/pull/4

File: src/battle-scene.ts [hunk 6]
```diff
@@ -160,6 +170,7 @@ export default class BattleScene extends SceneBase {
   public arenaNextEnemy: ArenaBase;
   public arena: Arena;
   public gameMode: GameMode;
+  public evilTeamThisRun: EvilTeam;
```
Comment (Xavion3) [inline src/battle-scene.ts:173]: Why is this being stored instead of being dynamically generated?

**Review by Xavion3 — COMMENTED** [review]
I'm still unsure on a lot of these teams, plentiful boss pokemon, and megas and paradox mixed in make me cautious as these feel like they're trying to compete with the rival and have more stuff going for them than the e4/champ. I feel like these boss teams could be pretty much universally toned down, and save these stronger teams for a hypothetical hard mode.

File: src/battle-scene.ts (hunk 6 above)
Comment (Tempo-anon) [reply src/battle-scene.ts:173]: Not sure what you mean by this. I dynamically generate it on line 828 `this.evilTeamThisRun = possibleEvilTeams[Utils.randInt(possibleEvilTeams.length)];` and then store it so the run generates grunts/bosses that match

Comment (Tempo-anon) [reply src/battle-scene.ts]: I ended up just changing how they generated completely and basically copied how the E4 is generated

File: src/battle.ts [hunk 7]
```diff
@@ -390,21 +341,27 @@ export const fixedBattles: FixedBattleConfigs = {
...
-  [113]: new FixedBattleConfig().setBattleType(BattleType.TRAINER)
-    .setGetTrainerFunc(scene => new Trainer(scene, getEvilTeamBattle(scene.evilTeamThisRun, "admin"), TrainerVariant.DEFAULT)),
-  [115]: new FixedBattleConfig().setBattleType(BattleType.TRAINER)
-    .setGetTrainerFunc(scene => new Trainer(scene, getEvilTeamBattle(scene.evilTeamThisRun, "boss1"), TrainerVariant.DEFAULT)),
+  [121]: new FixedBattleConfig().setBattleType(BattleType.TRAINER).setSeedOffsetWave(35)
+    .setGetTrainerFunc(getRandomTrainerFunc([ TrainerType.ROCKET_GRUNT, TrainerType.MAGMA_GRUNT, TrainerType.AQUA_GRUNT, TrainerType.GALACTIC_GRUNT, TrainerType.PLASMA_GRUNT, TrainerType.FLARE_GRUNT ])),
+  [123]: new FixedBattleConfig().setBattleType(BattleType.TRAINER).setSeedOffsetWave(35)
+    .setGetTrainerFunc(getRandomTrainerFunc([ TrainerType.ROCKET_GRUNT, TrainerType.MAGMA_GRUNT, TrainerType.AQUA_GRUNT, TrainerType.GALACTIC_GRUNT, TrainerType.PLASMA_GRUNT, TrainerType.FLARE_GRUNT ])),
+  [125]: new FixedBattleConfig().setBattleType(BattleType.TRAINER).setSeedOffsetWave(35)
```
Comment (flx-sta) [inline src/battle.ts:359]: I’ve never seen this `  [125]: `.
Would you have a sauce/explanation for me ?

Comment (Tempo-anon) [reply src/battle.ts:359]: That means on wave 125 it will generate a battle based off the seed of wave 35 where each boss_2 lines up with the corresponding team grunt generated on wave 35

Comment (flx-sta) [reply src/battle.ts:359]: Okay. But how is that legit code. I’ve never seen that syntax before 👀 or is it basically like
```ts
[
 “Key”: “value”,
 [150]: “value”
]
```
?

**Comment by FogMP** [comment]
I've been testing this feature recently, at different points of account progression(new account/only egg moves allowed/no restrictions) and it makes classic much more interesting to go through. These new trainers are added at points were not much really happen in classic, and they scale pretty well difficulty wise (The first evil team fight at 115 introduces boss in trainer battles, next rival fight has 2 bosses and finally the 165 evil team has 3), and this kind of difficulty was needed in my opinion.

It changes how you go through classic because now not only you have to consider which pokemon to bring just for the rival and the final boss, but also once you see which evil team you get at wave 35, you can start making a team that counters these fights.

Something that could be in consideration is that some of these evil teams use paradox pokemon, which are treated differently in pokerogue and you don't really meet at all until you are near to the end of a run, realistically this isn't that big of a deal, but to make classic more interesting the evil team fights could be added after you beat classic once, not only to make it more challenging but to give more variety to the gamemode  (which it really needs).

File: src/battle.ts (hunk 7 above)
Comment (Tempo-anon) [reply src/battle.ts:359]: It uses this interface
```ts
interface FixedBattleConfigs {
    [key: integer]: FixedBattleConfig
}
```

File: src/locales/de/dialogue.ts [hunk 8]
```diff
@@ -367,6 +367,201 @@ export const PGMdialogue: DialogueTranslationEntries = {
...
+    "encounter": {
+      1: "Keiner, der sich Team Magma in den Weg stellt, bekommt Gnade, nicht einmal Kinder!"
+    },
+    "victory": {
+      1: "Wie kann das sein? Ich bin Teil vom mächtigen Team Magma! Wir wollen doch nur die Welt verbessern…"
+    },
+  },
+  "aqua_grunt": {
+    "encounter": {
+      1: "Du willst dich also mit Team Aqua anlegen? Du traust dich ja was…Dich werfe ich über Bord!",
```
Comment (CodeTappert) [inline src/locales/de/dialogue.ts:391]: Is missing a space before the "Dich werfe". 

Could you add that? Dont want to create a whole new PR for that

File: src/locales/de/dialogue.ts [hunk 9]
```diff
@@ -367,18 +367,201 @@ export const PGMdialogue: DialogueTranslationEntries = {
       1: "Ich werde für das nächste Rennen tunen."
     },
   },
-  "firebreather": {
+  "rocket_grunt": {
     "encounter": {
-      1: "My flames shall devour you!",
-      2: "My soul is on fire. I'll show you how hot it burns!",
-      3: "Step right up and take a look!"
+      1: `Jetzt gibts Ärger!…
```
Comment (FredeX) [inline src/locales/de/dialogue.ts:372]: "gibt's" would be correct

File: src/locales/de/dialogue.ts [hunk 10]
```diff
@@ -367,18 +367,201 @@ export const PGMdialogue: DialogueTranslationEntries = {
...
+    "victory": {
+      1: "Interessant. Und ziemlich eigenartig."
+    },
+    "defeat": {
+      1: "Ich werde meine neue Welt erschaffen..."
+    }
+  },
+  "galactic_boss_cyrus_2": {
+    "encounter": {
+      1: `So sehen wir uns wieder. Es scheint, als wären unsere Schicksale miteinander verflochten
```
Comment (FredeX) [inline src/locales/de/dialogue.ts:509]: Missing a dot after "verflochten", right?

File: src/locales/de/dialogue.ts (hunk 9 above)
Comment (CodeTappert) [reply src/locales/de/dialogue.ts:372]: No. The pokewiki has it as "gibts" https://www.pokewiki.de/Team_Rocket_(Anime)#Mottos

File: src/locales/de/dialogue.ts (hunk 10 above)
Comment (CodeTappert) [reply src/locales/de/dialogue.ts:509]: yes

File: src/locales/de/dialogue.ts (hunk 9 above)
Comment (FredeX) [reply src/locales/de/dialogue.ts:372]: either "gibt's" (season 9) or "gibt es" but "gibts" is just wrong.

Comment (CodeTappert) [inline src/locales/de/dialogue.ts:372]:
```suggestion
      1: `Jetzt gibt es Ärger…
```

Comment (FredeX) [reply src/locales/de/dialogue.ts:372]: and if we follow the wiki there is no exclamation mark after Jessies first sentence

**Review by FredeX — APPROVED** [review]

File: src/locales/de/dialogue.ts (hunk 9 above)
Comment (CodeTappert) [reply src/locales/de/dialogue.ts:372]: See https://github.com/pagefaultgames/pokerogue/pull/1577#discussion_r1624271960 

I requested a change there.

Season 9 has a completly different Motto

Comment (FredeX) [reply src/locales/de/dialogue.ts]: alright whatever, good fix.

**Comment by CodeTappert** [comment]
Hey i just found https://www.deviantart.com/akuma-tsubasa/art/Sprite-Commission-LightningKillua15-4-843105484 which are free to use (when credited) sprites of Cipher.

Could this be something we add in this pr? If not then i will just do a PR after yours is merged. 

If you want to add this to this one here then i will provide you the solo sprite images and json and all

**Comment by Tempo-anon** [comment]
> Hey i just found https://www.deviantart.com/akuma-tsubasa/art/Sprite-Commission-LightningKillua15-4-843105484 which are free to use (when credited) sprites of Cipher.
> 
> Could this be something we add in this pr? If not then i will just do a PR after yours is merged.
> 
> If you want to add this to this one here then i will provide you the solo sprite images and json and all

I haven't really thought about Cipher since they're from a spin-off. You're more than welcome to make your own PR to add them once this is in though

**Comment by Tempo-anon** [comment]
Added evil team assets (sprites, music, teams, dialogue). They do **NOT** spawn so there is no gameplay impact whatsoever.

**Review by brain-frog — APPROVED** [review]
